docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,446 @@
+
+HelmChartProxy can use an approach similar to templated variables in ClusterClass to fill in variables values. The key/value map can include Go templating representing fields from the Cluster definition, i.e. cluster name, namespace, control plane ref, pod CIDRs. These template fields will be resolved dynamically based on the actual values for each workload cluster selected by HelmChartProxy’s cluster selector.
+
+In a future iteration we might consider to leverage template variables in ClusterClass. If a HelmChartProxy is selecting clusters with a managed topology, then the Go template configuration can benefit from Cluster.spec.topology.variables, and from the strong typing/validation ensured by the variable schema defined in corresponding ClusterClass.
+
+#### Maintain an inventory of add-ons installed in a Cluster
+
+Maintaining an inventory of add-ons installed is useful for a number of reasons. The most straightforward is for users to see at a glance which add-ons are installed on a cluster. On the other hand, the inventory is useful for add-on orchestration as well. It allows the controller to be able to distinguish between an add-on installed through the orchestrator and an add-on installed out of band. This means that on add-on deletion, the controller will only uninstall add-ons it is responsible for and won’t remove existing add-ons.
+
+First, the simplest is to let the package manager maintain the list. For example, `helm list` lists all the releases on a cluster which already handles maintaining an inventory of addons. However, there is no way to distinguish between out of band add-ons and add-ons installed through the orchestrator. Additionally, any ClusterAddonProvider would need to rely on their package manager to have a suitable implementation of listing packages on a cluster.
+
+The solution implemented in the ClusterAddonProvider for Helm is to use HelmReleaseProxy to maintain an inventory (see prototype design for more specifics). Each HelmReleaseProxy corresponds to a Helm release installed through HelmChartProxy. The controller will only orchestrate Helm releases with a corresponding HelmReleaseProxy, ensuring that out of band Helm releases won’t be affected. 
+
+Additionally, each HelmReleaseProxy will include a label indicating the Cluster the release is installed on, and an additional label indicating the HelmChartProxy it corresponds to. This provides a way to query for the Helm releases installed on a Cluster and also the Helm releases installed from a HelmChartProxy. This implementation also generalizes to any ClusterAddonProvider as it uses a CRD to maintain the inventory instead of the package manager.
+
+#### Surface Cluster Add-ons Status
+
+The add-on orchestrator is responsible for installing add-ons on an arbitrary number of workload clusters. As such, it would be beneficial to see the status of an add-on from the management cluster and avoid having to fetch the kubeconfig of the workload cluster.
```

Comment:
is "add-on orchestrator" the same as ClusterAddonProvider?

Reply:
Yep, I can reword that